### PR TITLE
1319927: [RFE] sub-man automatically enables yum plugins

### PR DESCRIPTION
**Pull request by jirihnidek**
* BZ [RFE]: https://bugzilla.redhat.com/show_bug.cgi?id=1319927
* Enable automatic enabling of yum plugins for sub-man CLI
* Automatic enabling of yum plugin is also implemented for sub-man GUI
* It is not implemented for initial-setup, because case desctibed
  in BZ bug report will be covered by sub-man CLI or GUI
* Add some unit tests for this case
* Added template to rhsm.conf file

### CODE REVIEW

**Comment by jirihnidek** [comment]
[1319927 - [RFE] subscription-manager should automatically enable yum plugins](https://trello.com/c/4L8cwRrk/533-1319927-rfe-subscription-manager-should-automatically-enable-yum-plugins)

File: src/subscription_manager/repolib.py [hunk 1]
```diff
@@ -60,13 +60,98 @@ def manage_repos_enabled():
...
+    def enable_yum_plugins(cls):
+        """
+        This function tries to enable yum plugins: subscription-manager and product-id.
+        It takes no action, when automatic enabling of yum plugins is disabled in rhsm.conf.
+        :return: It returns list of enabled plugins
+        """
+
+        # When user doesn't want to automatically enable yum plugins, then return empty list
+        if cls.is_auto_enable_enabled() is False:
+            log.info('The rhsm.auto_enable_yum_plugins is disabled')
```
Comment (barnabycourt) [inline src/subscription_manager/repolib.py]: Please add more context about what action is happening or not happening. For example: "The rhsm.auto_enable_yum_plugins is disabled. Skipping the enablement of yum plugins" or something like that.

Comment (jirihnidek) [reply src/subscription_manager/repolib.py]: I agree.

File: src/subscription_manager/repolib.py [hunk 2]
```diff
@@ -60,13 +60,98 @@ def manage_repos_enabled():
...
+                log.debug('Yum plugin: "%s" already enabled. Nothing to do.' % yum_plugin_file_name)
+            elif is_plugin_enabled == cls.YUM_PLUGIN_DISABLED:
+                log.warn('Enabling yum plugin: "%s".' % yum_plugin_file_name)
+                # Change content of plugin configuration file and enable this plugin.
+                with open(yum_plugin_file_name, 'w') as cfg_file:
+                    yum_plugin_config.set('main', 'enabled', cls.YUM_PLUGIN_ENABLED)
+                    yum_plugin_config.write(cfg_file)
+                enabled_yum_plugins.append(yum_plugin_file_name)
+            else:
+                log.warn('Value of rhsm.auto_enable_yum_plugins: "%s" is invalid.' % is_plugin_enabled)
```
Comment (cnsnyder) [inline src/subscription_manager/repolib.py]: Is it the rhsm.auto_enabled_yum_plugins value that is invalid? Or is it main.enabled (for yum_plugin_file_name)?

I think this should be changed to `"Invalid value ('%s') found in yum plugin configuration file '%s'" % (yum_plugin_file_name, is_plugin_enabled)`

Or something similar

File: test/test_repolib.py [hunk 3]
```diff
@@ -880,3 +884,114 @@ def test_empty_manage_repos_int_37(self):
...
+[rhsm]
+auto_enable_yum_plugins = 0
+"""
+
+YUM_PLUGIN_CONF_FILE_ENABLED = """
+[main]
+enabled = 1
+"""
+
+YUM_PLUGIN_CONF_FILE_DISABLED = """
```
Comment (cnsnyder) [inline test/test_repolib.py]: Perhaps add a test case for a plugin configuration file that does not match expected values.
`YUM_PLUGIN_CONF_FILE_INVALID = """
[main]
enabled = 4
"""

File: src/subscription_manager/repolib.py (hunk 2 above)
Comment (jirihnidek) [reply src/subscription_manager/repolib.py]: Yeah, you are right. This log message is wrong. There should be this: `"Invalid value ('%s') found in yum plugin configuration file '%s'" % (is_plugin_enabled, yum_plugin_file_name)"`

File: test/test_repolib.py (hunk 3 above)
Comment (jirihnidek) [reply test/test_repolib.py]: ACK.

File: src/subscription_manager/repolib.py [hunk 4]
```diff
@@ -60,13 +60,110 @@ def manage_repos_enabled():
...
+            if auto_enable_yum_plugins is None:
+                auto_enable_yum_plugins = True
+        return bool(auto_enable_yum_plugins)
+
+    @staticmethod
+    def warning_message(enabled_yum_plugins):
+        message = _('The yum plugins: %s were automatically enabled for the benefit of '
+                    'Red Hat Subscription Management. If not desired, use '
+                    '"subscription-manager config --rhsm.auto_enable_yum_plugins=0" to '
+                    'block this behavior.\n') % str(tuple(enabled_yum_plugins))
```
Comment (jsefler) [inline src/subscription_manager/repolib.py]: Can we use something different that tuple to join the enabled_yum_plugins?  The current possible results of this string are:
('/etc/yum/pluginconf.d/subscription-manager.conf',)
('/etc/yum/pluginconf.d/product-id.conf',)
('/etc/yum/pluginconf.d/subscription-manager.conf', '/etc/yum/pluginconf.d/product-id.conf')

Only the third one looks well punctuated.

**Pull request merged**